### ISSUE TITLE
CAS error: Provided service did not match ticket service

### ISSUE THREAD

**Issue by anamba** [uhm-coe/authorizer#57]
The latest version (2.8.1) is not working for me, I get messages like this in the php error log:
```
Authentication failure: Ticket not validated\nReason: [INVALID_SERVICE] CAS error: Provided service did not match ticket service
```

I rolled back to 2.8.0, which still didn't work, and finally to 2.7.2, which did work. So that is what I am doing for now.

However, I did want to let you know my findings, so that hopefully we can get back to running on the latest version again.

Looking closely at the CAS Server access log, I noticed that the service URL appears to be truncated:
```
GET /cas/p3/serviceValidate?service=https%3A%2F%2F[my.domain]%2Fwp-login.php%3Fexternal%3Dcas%26redirect_to%3Dhttps%253A&ticket=ST-Nv0c327vLjSOLgh4G2KIRFzaFv75klLLOp72-zb_B2E
```

I suspected that might be the cause of the mismatch, so I looked for references to that in the code. I found that if I comment out line 1300, CAS login magically works again:
```
if ( isset( $login_querystring['redirect_to'] ) ) {
  #$cas_service_url .= '&redirect_to=' . rawurlencode( $login_querystring['redirect_to'] );
}
```

Any thoughts?

**Comment by pkarjala** [comment]
Thank you @anamba for the bug report!  We'll start looking at this and report back once we have an update.

**Comment by pkarjala** [comment]
Hi @anamba looking into this issue, we are not yet able to reproduce it.  Can you provide some details about your environment?

- PHP Version
- WordPress Version
- Other WordPress plugins installed  (it is recommended to try disabling other plugins to test for conflicts with Authorizer functionality)

Additionally, does this error occur when you go directly to https://yourdomain.com/wp-login.php and press the Authorizer login button?

Finally, can you check that the time settings on your WordPress server and your CAS server are in sync?  If there is a discrepancy beyond a few minutes, it can potentially cause issues.

Looking at the redirect_to value in your server log, it looks like it's improperly encoding the value `https:` which is being passed already encoded as `https%3A` and then being double-encoded as `https%253A`.  The function `rawurlencode` doesn't truncate when it creates the redirection URL.

Can you try replacing lines 1296 - 1301 with the following, then check your logs when you try to log in using the CAS button?  This should output the values being passed in, and hopefully give some clues as to what's going on when using Authorizer 2.8.1.

```
			if ( isset( $_SERVER['QUERY_STRING'] ) ) {
				error_log( '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~' );
				error_log( 'Raw Query string is: ' );
				error_log( print_r( $_SERVER['QUERY_STRING'], true ) );
				parse_str( wp_parse_url( esc_url_raw( wp_unslash( $_SERVER['QUERY_STRING'] ) ), PHP_URL_HOST ), $login_querystring );
				error_log( 'New parsed querystring is: ' );
				error_log( print_r( $login_querystring, true ) );
				error_log( '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~' );
			}
			if ( isset( $login_querystring['redirect_to'] ) ) {
				error_log( '=======================================' );
				error_log( 'Raw login querystring value is:' );
				error_log( print_r( $login_querystring['redirect_to'], true ) );
				$cas_service_url .= '&redirect_to=' . rawurlencode( $login_querystring['redirect_to'] );
				error_log( 'New cas_service_url is: ' );
				error_log( print_r( $cas_service_url, true ) );
				error_log( '=======================================' );
			}
```

**Comment by figureone** [comment]
Thanks for the detailed report @anamba. Your `redirect_to` definitely looks like it's getting truncated by the time it reaches the CAS server, which may mess up the parsing of the CAS `ticket` param. It's basically leaving you with a double-encoded `https:` (`https:` => `https%3A` => `https%253A`) instead of the full double-encoded redirect URL.

If 2.7.2 works for you, the problem is likely due to a switch from `urlencode()` to `rawurlencode()`, added in 2.8.0 here: https://github.com/uhm-coe/authorizer/commit/063a5fa26af2250f002c19a2f2cc203462b62c7c

The switch was due to following WordPress Coding Standards, in particular, this rule:

> urlencode() should only be used when dealing with legacy applications rawurlencode() should now be used instead. See http://php.net/manual/en/function.rawurlencode.php and http://www.faqs.org/rfcs/rfc3986.html (WordPress.PHP.DiscouragedPHPFunctions.urlencode_urlencode)

There's not supposed to be big differences between the two functions, other than `urlencode()` encoding spaces as `+` and `rawurlencode()` using `%20`, and `rawurlencode()` encoding `~` on PHP versions less than 5.3.
https://stackoverflow.com/questions/996139/urlencode-vs-rawurlencode

If you're willing, you can modify @pkarjala's debug code above to see if `rawurlencode()` and `urlencode()` return different values on your system, and we can go from there. Thanks!

**Comment by figureone** [comment]
Thanks for the extra details! Looks like the problem is related to a refactor a few lines up, in the `parse_str()` call, introduced in this pull request:
https://github.com/uhm-coe/authorizer/commit/fffaa6d9244284e7e6c46aadf317a0d260b1a6a4

We'll investigate further and see if we can narrow it down.

**Comment by anamba** [comment]
Yes, can confirm that removing `wp_parse_url` and `esc_raw_url` makes it work again with the `redirect_to` param.

The web server config isn't anything too unusual, I don't think... just php-fpm (of the two sites I tested, one was 5.6 and the other was 7.1) behind apache 2.4 + mod_proxy_fcgi.

**Comment by pkarjala** [comment]
Great!  We'll likely roll out the change soon.  Thank you for the help in testing!

**Comment by figureone** [comment]
Just released version 2.8.2 with the fix: https://github.com/uhm-coe/authorizer/commit/00ca21991604db31c2db1faae2cc9f8eedcee8b0

Thanks for you help debugging this one!
